torchci/pages/api/drci/drci.ts

```diff
@@ -41,25 +52,121 @@ export async function updateDrciComments(octokit: Octokit, prNumber?: string) {
     );
 
     const workflowsByPR = reorganizeWorkflows(recentWorkflows);
+    await addMergeBaseCommits(octokit, workflowsByPR);
     const sevs = getActiveSEVs(await fetchIssuesByLabel("ci: sev"));
+    const masterFlakyJobs = await getFlakyJobs();
+    const baseCommitJobs = await getBaseCommitJobs(workflowsByPR);
 
-    for (const [pr_number, pr_info] of workflowsByPR) {
-        const { pending, failedJobs } = getWorkflowJobsStatuses(pr_info);
+    await Promise.all(
+      Array.from(workflowsByPR.values()).map(async (pr_info) => {
+        const { pending, failedJobs, flakyJobs, brokenTrunkJobs } =
+          getWorkflowJobsStatuses(
+            pr_info,
+            masterFlakyJobs,
+            baseCommitJobs.get(pr_info.merge_base) || new Map()
+          );
 
-        const failureInfo = constructResultsComment(pending, failedJobs, pr_info.head_sha);
-        const comment = formDrciComment(pr_number, failureInfo, formDrciSevBody(sevs));
+        const failureInfo = constructResultsComment(
+          pending,
+          failedJobs,
+          flakyJobs,
+          brokenTrunkJobs,
+          pr_info.head_sha,
+          pr_info.merge_base
+        );
+        const comment = formDrciComment(
+          pr_info.pr_number,
+          failureInfo,
+          formDrciSevBody(sevs)
+        );
 
         await updateCommentWithWorkflow(octokit, pr_info, comment);
+      })
+    );
+}
+
+async function addMergeBaseCommits(
+  octokit: Octokit,
+  workflowsByPR: Map<number, PRandJobs>
+) {
+  await Promise.all(
+    Array.from(workflowsByPR.values()).map(async (pr_info) => {
+      const diff = await octokit.rest.repos.compareCommits({
+        owner: OWNER,
+        repo: REPO,
+        base: pr_info.head_sha,
+        head: "master",
+      });
+      pr_info.merge_base = diff.data.merge_base_commit.sha;
+    })
+  );
+}
+
+async function getBaseCommitJobs(
+  workflowsByPR: Map<number, PRandJobs>
+): Promise<Map<string, Map<string, RecentWorkflowsData>>> {
+  // get merge base shas
+  let baseShas = [];
+  for (const [_, pr_info] of workflowsByPR) {
+    baseShas.push(pr_info.merge_base);
+  }
+
+  // fetch failing jobs on those shas
+  const commitFailedJobsQueryResult = await fetchFailedJobsFromCommits(baseShas);
+
+  // reorganize into a map of sha -> name -> data
+  const jobsBySha = new Map();
+  for (const job of commitFailedJobsQueryResult) {
+    if (!jobsBySha.has(job.head_sha)) {
+      jobsBySha.set(job.head_sha, new Map());
+    }
+    const existing_job = jobsBySha.get(job.head_sha).get(job.name!);
+    if (!existing_job || existing_job.id < job.id!) {
+      // if rerun, choose the job with the larger id as that is more recent
+      jobsBySha.get(job.head_sha).set(job.name, job);
     }
+  }
+  return jobsBySha;
+}
+
+async function getFlakyJobs(): Promise<FlakyJob[]> {
```

Comment:
Should we have a generic `fetchJSON` method (or just use it here rather than implement fetch + parse logic again?)

Reply:
done